types/jsforce/record.d.ts

```diff
@@ -1,6 +1,16 @@
+import { RecordResult } from './record-result';
+import { Connection } from './connection';
 import { SalesforceId } from './salesforce-id';
+import { Stream } from 'stream';
 
-export interface Record {
-    Id: SalesforceId;
-    attributes: Object[];
+export class RecordReference<T> {
+    constructor(conn: Connection, type: string, id: SalesforceId);
+    blob(fieldName: string): Stream;
+    del(options?: Object, callback?: (err: Error, result: RecordResult) => void): Promise<RecordResult>;
+    delete(options?: Object, callback?: (err: Error, result: RecordResult) => void): Promise<RecordResult>;
+    destroy(options?: Object, callback?: (err: Error, result: RecordResult) => void): Promise<RecordResult>;
+    retrieve(options?: Object, callback?: (err: Error, record: Record<T>) => void): Promise<Record<T>>;
+    update(record: Partial<T>, options?: Object, callback?: (err: Error, result: RecordResult) => void): Promise<RecordResult>;
 }
+
+export type Record<T> = {Id: SalesforceId } & T;
```

Comment:
A Record is the actual data in salesforce, so all we can say about it generically is that it has an Id.  Any other properties need to be defined in a data type and inferred.  You can always do `Record<any>` to escape the compiler.